Add opacity control to MouseRegion. Add findAnnotations to Layer.

dkwingsmt opened the pull request:
## Description
This PR adds a new parameter `bool opaque` to `MouseRegion`. In order to do this, the same parameter is added to `RenderMouseRegion` and `AnnotatedRegionLayer`, and a new protected method `findAnnotations` is added to `Layer` to unify the implementation of `find` and `findAll`. 

For detailed explanation of changes and migration, see https://github.com/flutter/flutter/issues/38488.

## Related Issues

- Detail, migration and discussion: https://github.com/flutter/flutter/issues/38488

## Tests

I added the following tests:
- `ContainerLayer` should correctly handle `findAll` and `find` under various situations.
- `AnnotatedRegionLayer` should correctly add annotations under various situations.
- For each of the other layer class in `layer.dart`, `findAll` should work property.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). I'll send out breaking change proposal we've passed the draft phase.
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

## Todo 
No test is added to `FollowerLayer` yet, because I suspect there are bugs related to it, and testing it requires non-trivial effort. I'll investigate it after this PR.

dkwingsmt committed:
Split MouseRegion out and fix tests

dkwingsmt committed:
Renames

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into split-mouse-listener

dkwingsmt committed:
Move test to deprecated test for ignore

dkwingsmt committed:
Fix deprecation and lint

dkwingsmt committed:
Rename Mouse -> MouseRegion

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into split-mouse-listener

dkwingsmt committed:
Make deprecation a TODO

dkwingsmt committed:
Fix sample code

dkwingsmt committed:
Fix deprecation message

dkwingsmt committed:
Rephrase TODO

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into split-mouse-listener

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into split-mouse-listener

dkwingsmt committed:
Fix unnecessary expectation that broke tests

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into split-mouse-listener

dkwingsmt committed:
Fix a few docs

dkwingsmt committed:
Fix docs

dkwingsmt committed:
Stop using static values

dkwingsmt committed:
Changing expectation to adapt with the change

dkwingsmt committed:
Annotation test

dkwingsmt committed:
Doc proof reading

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into split-mouse-listener

dkwingsmt committed:
Fix doc because mouse region allows buttons

dkwingsmt committed:
Empty line

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into split-mouse-listener

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into split-mouse-listener

dkwingsmt committed:
Rename to RenderMouseRegion

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Reimplement hitTest; findAll uses hitTest.

dkwingsmt committed:
Fix compile

dkwingsmt committed:
Add behavior to MouseRegion

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Fix compile. Add docs.

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Add some tests

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Test MouseRegion.opaque

dkwingsmt committed:
Test default value

dkwingsmt committed:
Fix lint

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Listener's MouseRegion keeps opaque false

dkwingsmt committed:
Improve deprecation message and docs

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -14,6 +14,46 @@ import 'package:vector_math/vector_math_64.dart';
+  final T value;
+
+  @override
+  String toString() => '$value';
```
prefer `'$runtimeType($value)'` otherwise it won't be clear what you're printing

dkwingsmt replied:
I agree. Although `HitTestEntry` is doing the same https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/gestures/hit_test.dart#L57

Hixie replied:
we should fix that one too

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -14,6 +14,46 @@ import 'package:vector_math/vector_math_64.dart';
+  @override
+  String toString() => 'LayerHitTestResult(${_path.isEmpty ? "<empty path>" : _path.join(", ")})';
+}
+
```
do we have to use new classes here, as opposed to reusing the equivalent classes from the gestures library?

dkwingsmt replied:
I'd love to reuse classes, but `handleEvent` doesn't make sense in layers https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/gestures/hit_test.dart#L42

Hixie replied:
fix that one too!

Hixie replied:
maybe we should adjust the gestures library objects so that they do make sense... not sure exactly how to do it

dkwingsmt replied:
We can also just use a `List<S>` here, since we only care about one list of values. Although it's not as future-proof.

dkwingsmt replied:
After discussion we discovered that, the mouse events also need `localPosition`, which require that extra classes are needed to be the return value. See https://github.com/flutter/flutter/issues/38488#issuecomment-527688254

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -150,7 +190,26 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// (preventing non-ancestral layers below this one from being hit).
+  /// Returns false if the hit can continue to other non-ancestral objects
+  /// behind this one.
+  bool hitTest<S>(LayerHitTestResult<S> result, { @required Offset regionOffset }) {
```
would be nice for this to be abstract, so that anyone creating a new layer is encouraged to implement it

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -617,18 +664,13 @@ class ContainerLayer extends Layer {
+        return true;
     }
-    return result;
+    return false;
```
do we intentionally not add this layer to the list as well?

dkwingsmt replied:
As far as it's designed right now, layer hit test only works for annotations. Therefore only `AnnotatedRegionLayer` can add anything to the list.
Also it is not the layers that are added, but the value.

Hixie replied:
seems weird to have hit testing work so differently in layers than in render objects.

dkwingsmt replied:
Honestly I was about to name it `findAnnotations` since their implementations are so different, but I couldn't think of a shorter name.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -150,7 +190,26 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   ///
   ///  * [AnnotatedRegionLayer], for placing values in the layer tree.
-  Iterable<S> findAll<S>(Offset regionOffset);
+  Iterable<S> findAll<S>(Offset regionOffset) {
```
`regionOffset` seems like a weird parameter name for this.

dkwingsmt replied:
Would you like me to rename all occurrences of `regionOffset` to something like `localPosition` or `position`?

Hixie replied:
sure

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -150,7 +190,26 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   ///
   ///  * [AnnotatedRegionLayer], for placing values in the layer tree.
-  Iterable<S> findAll<S>(Offset regionOffset);
+  Iterable<S> findAll<S>(Offset regionOffset) {
```
seems weird that findAll and find are not diverged in implementation so much. could one be implemented in terms of the other?

jonahwilliams replied:
Originally find was re-implemented in terms of findAll in @gspencergoog 's original PR, but that end up regressing several benchmarks. The performance might have been fixed since then

Hixie replied:
i was thinking more the other way around (findAll is by design more expensive than find)

dkwingsmt replied:
I thought for a very long time and see if there was a way that could make `find` and `findAll` share code with methods such as `visitChildren` and `visitSelf`. I could not come up with a way without introducing several ridiculously complex callbacks. I've decided to add another parameter `bool onlyFirst`.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2038,9 +2084,14 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
+    this.opaque = false,
+  }) : assert(value != null),
+       assert(opaque != null),
+       offset = offset ?? Offset.zero;
```
do we intentionally accept `null` as a value for `offset`?

Hixie replied:
(as opposed to `Offset.zero` being the default value)

gspencergoog replied:
When forwarding values (from subclass constructors, for instance), it often happens that just passing on a parameter that is missing converts it to a null.  It means that the same default needs to be applied in a lot of places,  instead of just in one, so this code is just guarding against that, and supplying the one default. And without this, if the default changes in one place, we need to remember to change it everywhere, which is error-prone.

dkwingsmt replied:
I agree.

And default values are already causing troubles, such as in https://github.com/flutter/flutter/pull/37896/files/daadb7b58e71117c5df325b9e1db70505f72beea#diff-96bb73b252a2e7c4b88089358c1ffd35R709 where I want to test the default value while reusing the test widget.

Hixie replied:
in general we prefer to update the defaults everywhere and to be careful. defaults are self-documenting, and more idiomatic. We should use tests to verify that we haven't missed cases when we change them.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2062,6 +2113,21 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
   final Offset offset;
 
+  /// Whether this layer should prevent layers visually behind it from receiving
+  /// the hit testing.
```
`receiving the hit testing` -> `being visible to hit tests` or some such

Hixie replied:
or "being hit tested"

dkwingsmt replied:
I changed it to `being to the hit during a hit test.`

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2062,6 +2113,21 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
+  /// Whether this layer should prevent layers visually behind it from receiving
+  /// the hit testing.
+  ///
+  /// If [opaque] is true, it will "absorb" the hit test if any of its children
```
"the hit test" -> "hits" (and "it" -> "them" on the next line)

dkwingsmt replied:
I changed it to "the hit". Describing the event during a single hit test might be easier to understand?

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2062,6 +2113,21 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
+  ///
+  /// If [opaque] is true, it will "absorb" the hit test if any of its children
+  /// absorb it, or if this layer contains the target position and has the
+  /// requested annotation. Absorbing the hit test prevents layers visually
```
`layers` -> `sibling layers (or other layers that are not ancestors or descendants of this layer)`

dkwingsmt replied:
I changed it to "A hit that is absorbed will not be received by this layer's siblings (or any other layers that are not ancestors or descendants of this layer)."

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2062,6 +2113,21 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
+  /// it to receive the test no matter whether the layer or its children would
+  /// absorb the test otherwise.
+  ///
+  /// This defaults to false.
```
```
///
/// See also:
///
///  * [HitTestBehavior], which controls similar logic when hit-testing in the render tree.
```

Hixie commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5533,11 +5533,11 @@ class Listener extends StatelessWidget {
     // and Listener should no longer need an intermediate class _PointerListener.
     // https://github.com/flutter/flutter/issues/36085
-    @Deprecated('Use MouseRegion.onEnter instead')
+    @Deprecated('Use MouseRegion.onEnter instead. Read MouseRegion.opaque for behavioral difference.')
```
Read -> See

Hixie commented on packages/flutter/test/rendering/layers_test.dart:
```diff
@@ -465,4 +465,213 @@ void main() {
+        expect(pathToValues(result.path), <int>[1]);
+      });
+    });
+  });
```
minor thing, but I would encourage you to not use `group` unless you have to. We've ended up with annoying bugs when we start using `group`s in more complicated ways, so we're generally moving away from them.

dkwingsmt committed:
Make hitTest abstract

dkwingsmt committed:
Fix doc

dkwingsmt committed:
Even better doc

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2062,6 +2123,26 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
+  /// Whether this layer should prevent layers visually behind it from being
+  /// hit during a hit test.
+  ///
+  /// If [opaque] is true, this layer will "absorb" the hit if any of its
```
```suggestion
  /// If [opaque] is true, this layer will absorb the hit if any of its
```

gspencergoog commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -2577,6 +2585,20 @@ class RenderMouseRegion extends RenderProxyBox {
+  /// from detecting the pointer, thus affecting how their [onHover], [onEnter],
+  /// and [onExit] behave.
+  ///
+  /// If [opaque] is true, this object will "absorb" the mouse pointer so that
```
```suggestion
  /// If [opaque] is true, this object will absorb the mouse pointer so that
```

gspencergoog commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5796,6 +5799,20 @@ class MouseRegion extends SingleChildRenderObjectWidget {
+  /// behind it from detecting the pointer, therefore affecting how their
+  /// [onHover], [onEnter], and [onExit] behave.
+  ///
+  /// If [opaque] is true, this object will "absorb" the mouse pointer so that
```
```suggestion
  /// If [opaque] is true, this object will absorb the mouse pointer so that
```

gspencergoog commented on packages/flutter/test/rendering/layers_test.dart:
```diff
@@ -465,4 +465,213 @@ void main() {
+
+      test('should respect size and offset, and always return children', () {
+
+        // Tests:
```
Nice ASCII art!

dkwingsmt committed:
doc: remove quotes

dkwingsmt committed:
Docs for MouseRegion and Render

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Implement hitTest

dkwingsmt committed:
Remove test grouping

dkwingsmt committed:
HitTestEntry.toStrong now includes runtimeTime

dkwingsmt committed:
Doc: findAll mentions hitTest

dkwingsmt committed:
Simplify changes by removing two classes

dkwingsmt committed:
Add onlyFirst to hitTest

dkwingsmt committed:
Opaque no longer flips absorption from children

dkwingsmt committed:
Rename hitTest to findAnnotations

dkwingsmt committed:
Rewrite layer_annotation_test

dkwingsmt committed:
Add tests for almost all layers

dkwingsmt committed:
Remove tests for FollowerLayer for now

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Remove redundant find

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -213,31 +213,70 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   ///
   /// Returns null if no matching region is found.
   ///
+  /// By default it calls [findAnnotations] with `onlyFirst: true` and returns
```
it -> this method

Hixie commented:
General approach seems reasonable. @goderbauer would you mind taking over to do a more detailed review?

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Improve doc

dkwingsmt committed:
regionOffset -> localPosition

dkwingsmt committed:
Change tooltip to using method

This is a leftover change from https://github.com/flutter/flutter/pull/36217#discussion_r306950227

dkwingsmt committed:
Add opaque to debug strings

dkwingsmt committed:
Fix a few docs

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Use AnnotationEntry on layers

dkwingsmt committed:
Add AnnotationResult

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Doc

dkwingsmt committed:
Test annotation location

dkwingsmt committed:
Fix system chrome

dkwingsmt committed:
Fix tests

dkwingsmt commented:
Update: I'm waiting for review after some non-trivial changes since the last review.

Also since I'll be away for 2 weeks, I'm postponing the merge to Sep 24th.

dkwingsmt committed:
Fix doc

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Make equalToAnnotationResult private

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

gspencergoog commented on packages/flutter/lib/src/material/tooltip.dart:
```diff
@@ -400,8 +408,8 @@ class _TooltipState extends State<Tooltip> with SingleTickerProviderStateMixin {
       result = MouseRegion(
-        onEnter: (PointerEnterEvent event) => _showTooltip(),
-        onExit: (PointerExitEvent event) => _hideTooltip(),
+        onEnter: _handleEnter,
```
Are these wrappers necessary, given that `MouseRegion` also wraps them? I thought that was the reason to wrap them in `MouseRegion`, so that callers don't have to.

dkwingsmt replied:
Ok, I reverted this change. To make style consistent, I applied the reversed change onto https://github.com/flutter/flutter/pull/37896/files#diff-f1c1016f733aebaa60581bf989e36d17L1107

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -1991,37 +2047,31 @@ class FollowerLayer extends ContainerLayer {
     if (link.leader == null) {
-      return showWhenUnlinked ? super.findAll<S>(regionOffset - unlinkedOffset) : <S>[];
+      if (showWhenUnlinked)
+        return super.findAnnotations(result, localPosition - unlinkedOffset, onlyFirst: onlyFirst);
```
nit: You might add curly braces around this, to be consistent with the one below.

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2192,35 +2247,56 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
   /// Ignored if [size] is not set.
+  ///
+  /// Unlike [OffsetLayer], this parameter [offset] does not affect the layer's
+  /// children.
```
You might explain what it does affect, rather than just what it doesn't affect.

dkwingsmt replied:
The first line of the doc already explains "The [offset] is optionally used to translate the clip region for the hit-testing of [find] by [offset]." I added this line because the fact that this API resembles `OffsetLayer.offset` is a little misleading.

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2192,35 +2247,56 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
-      return childResults;
+  @protected
+  bool findAnnotations<S>(AnnotationResult<S> result, Offset localPosition, { @required bool onlyFirst }) {
+    bool isAbsorbed = super.findAnnotations(result, localPosition, onlyFirst: onlyFirst);
```
What's the relationship between `onlyFirst` and `opaque`? It seems like you should specify true for `onlyFirst` if `opaque` is true, to avoid searching further than you need to.  But maybe I'm just not understanding what the relationship is...

dkwingsmt replied:
Parameter `onlyFirst` is a property of the tree walking _process_(function call), while `opaque` is a property of _layers_.

If `onlyFirst` is true, then the layers' `opaque` doesn't actually matter.

If `onlyFirst` is not true, then `opaque` can change results. Consider the following case, where C is parent to A and B, and we call `C.findAnnotations` with the position of the pointer. 
```
       | Pointer
       v 
      ------- B
           |
A ------   |
   |       |
C------------
```
If B is opaque, then the result is `[B, C]`; if B is not opaque, then the result is `[B, A, C]`. (Apparently, if `onlyFirst` is true, then the result is always `[B]`.)

dkwingsmt replied:
I'll appreciate it if you can help me with the documentation to make it more comprehensible :)

gspencergoog replied:
Sure, no problem!

So, why isn't the result just `[B]` if B is opaque? I thought that was the point of it being opaque, that it stopped on the first hit.

dkwingsmt replied:
Opacity only controls the relationship between siblings. (Parents usually returns if any of their children returns true.)

In other words, if one layer is hit, then its ancestors are _always_ hit too.

(This algorithm is similar to `HitTestBehavior.opaque` in render object hit testing.)

gspencergoog commented on packages/flutter/test/rendering/layer_annotations_test.dart:
```diff
@@ -0,0 +1,758 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
Should be 2019, since this is a new file.

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Minor stuff for review

dkwingsmt committed:
Make code style consistent and concise

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

gspencergoog commented on packages/flutter/lib/src/gestures/hit_test.dart:
```diff
@@ -54,7 +54,7 @@ class HitTestEntry {
 
   @override
-  String toString() => '$target';
+  String toString() => '$runtimeType($target)';
```
You might use `'${describeIdentity(this)}($target)'` instead, since it will add a hash to the runtime type that lets you see if two objects are the same object or not.

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2192,35 +2248,56 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
+  /// the target position meets the requirement defined by [size] and [offset].
+  ///
+  /// The return value is true if any child returns true, or the layer's
+  /// annotation is added and [opaque] is true.
```
You should add a description of the `onlyFirst`, `result`, and `localPosition` parameters to these docs.  I like your description in the comment below, but the ASCII art won't work here.

gspencergoog replied:
For `onlyFirst`, maybe something like:

The [onlyFirst] parameter indicates that the while walking the layer tree, the search will stop at the first child at [localPosition]. If [onlyFirst] is true, then the value of [opaque] is ignored. If [onlyFirst] is false, then [opaque] is used to decide whether or not the tree walk continues to the other children.  If [opaque] is false, then [findAnnotations] will return true if there is any layer at [localPosition].

The parent will return true if any of the children return true from their [findAnnotations].

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2192,35 +2248,56 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
+  ///
+  ///  * [HitTestBehavior], which controls similar logic when hit-testing in the
+  ///    render tree.
+  final bool opaque;
```
You should describe how this is affected by the `onlyFirst` argument in `findAnnotations`.

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2192,35 +2248,56 @@ class AnnotatedRegionLayer<T> extends ContainerLayer {
-    }
-    return null;
-  }
+  /// Whether this layer should prevent layers visually behind it from being
```
You should turn this comment into a macro with `{@template ...}...{@endtemplate}` (see other macros for an example), and use the `{@macro ...}` on the other `opaque` attributes in this file.

dkwingsmt committed:
Add hash to HitTestEntry.toString

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Revamp docs

dkwingsmt committed:
Add relationship between opaque and onlyFirst.

dkwingsmt commented:
@gspencergoog I've rewritten most docs, PTAL
https://github.com/flutter/flutter/compare/316aab4...b2405d5

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -213,31 +276,105 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  ///
+  /// ## Parameters and return value
+  ///
+  /// The [result] parameter is where the method output the resultant
```
```suggestion
  /// The [result] parameter is where the method outputs the resulting
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -213,31 +276,105 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// ## Parameters and return value
+  ///
+  /// The [result] parameter is where the method output the resultant
+  /// annotations. New annotations found during the walk is added to the tail.
```
```suggestion
  /// annotations. New annotations found during the walk are added to the tail.
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -213,31 +276,105 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// annotations. New annotations found during the walk is added to the tail.
+  ///
+  /// The [onlyFirst] parameter indicates that, if true, the search will stop
+  /// when finding the first qualified annotation; otherwise, it will walk the
```
```suggestion
  /// when it finds the first qualified annotation; otherwise, it will walk the
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -213,31 +276,105 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// entire subtree.
+  ///
+  /// The return value indicates the opacity of this layer and its subtree at
+  /// this position. If it is true, then this layer's parent should skip the
```
```suggestion
  /// this position. If it returns true, then this layer's parent should skip the
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -213,31 +276,105 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// The return value does not affect whether the parent adds its own
+  /// annotations; in other words, if a layer is supposed to add an annotation,
+  /// it will always add it even if its children are opaque to this type of
+  /// annotation, after its children. However, the opacity that the parents
```
```suggestion
  /// annotation. However, the opacity that the parents
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2160,67 +2243,127 @@ class FollowerLayer extends ContainerLayer {
+/// A composited layer which annotates its children with a value. Pushing this
+/// layer to the tree is the common way of adding an annotation.
+///
+/// Annotation is an optional object of any type that, when carried with a
```
```suggestion
/// An annotation is an optional object of any type that, when attached to a
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2160,67 +2243,127 @@ class FollowerLayer extends ContainerLayer {
+/// Annotation is an optional object of any type that, when carried with a
+/// layer, can be retrieved using [Layer.find] or [Layer.findAll] with a
+/// position. The search process is done recursively, controlled by a concept
+/// of being opaque to a type of annotation, explained at the return value of
```
```suggestion
/// of being opaque to a type of annotation, explained in the documentation of
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2160,67 +2243,127 @@ class FollowerLayer extends ContainerLayer {
+/// of being opaque to a type of annotation, explained at the return value of
+/// [Layer.findAnnotations].
+///
+/// When an annotation search arrives, this layer first defers the same search
```
```suggestion
/// When an annotation search arrives, this layer defers the same search
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2160,67 +2243,127 @@ class FollowerLayer extends ContainerLayer {
+/// [Layer.findAnnotations].
+///
+/// When an annotation search arrives, this layer first defers the same search
+/// to each of this layer's child, respecting their opacity. Then it adds this
```
```suggestion
/// to each of this layer's children, respecting their opacity. Then it adds this
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2160,67 +2243,127 @@ class FollowerLayer extends ContainerLayer {
-/// These values can be retrieved using [Layer.find] with a given [Offset]. If
-/// a [Size] is provided to this layer, then find will check if the provided
-/// offset is within the bounds of the layer.
+/// This layer is opaque to a type of annotation if any child is so, or if
```
```suggestion
/// This layer is opaque to a type of annotation if any child is also opaque, or if
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2160,67 +2243,127 @@ class FollowerLayer extends ContainerLayer {
+  ///
+  /// If [opaque] is false, or if this layer does not add its annotation,
+  /// then the opacity of this layer will be the one returned by the children,
+  /// meaning that it is be opaque if any child is opaque.
```
```suggestion
  /// meaning that it will be opaque if any child is opaque.
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2160,67 +2243,127 @@ class FollowerLayer extends ContainerLayer {
+  ///
+  /// This method always searches its children, and if any child returns `true`,
+  /// the remaining children are skipped. Regardless of what the children
+  /// returns, this method then adds this layer's annotation if all of the
```
```suggestion
  /// return, this method then adds this layer's annotation if all of the
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2160,67 +2243,127 @@ class FollowerLayer extends ContainerLayer {
+  /// {@macro flutter.rendering.annotatedRegionLayer.restrictions}
+  ///
+  /// This search process respects `onlyFirst`, meaning that when `onlyFirst` is
+  /// true, the search will stop when it finds first annotation from the
```
```suggestion
  /// true, the search will stop when it finds the first annotation from the
```

gspencergoog commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -2160,67 +2243,127 @@ class FollowerLayer extends ContainerLayer {
+  /// This search process respects `onlyFirst`, meaning that when `onlyFirst` is
+  /// true, the search will stop when it finds first annotation from the
+  /// children, and the layer's own annotation is checked only when none is
+  /// given from the children.
```
```suggestion
  /// given by the children.
```

gspencergoog commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -2592,7 +2592,8 @@ class RenderPointerListener extends RenderProxyBoxWithHitTestBehavior {
 ///
 /// Simply put, it responds to events that are related to hovering,
-/// i.e. when the mouse enters, exits or hovers a region without pressing.
+/// i.e. when the mouse enters, exits or moves over a region (with or without
```
```suggestion
/// i.e. when the mouse enters, exits, or moves over a region (with or without
```

gspencergoog commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -2620,10 +2628,24 @@ class RenderMouseRegion extends RenderProxyBox {
+  /// from detecting the pointer, thus affecting how their [onHover], [onEnter],
+  /// and [onExit] behave.
+  ///
+  /// If [opaque] is true, this object will absorb the mouse pointer, and
```
```suggestion
  /// If [opaque] is true, this object will absorb the mouse pointer and
```

gspencergoog commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -2620,10 +2628,24 @@ class RenderMouseRegion extends RenderProxyBox {
+  /// and [onExit] behave.
+  ///
+  /// If [opaque] is true, this object will absorb the mouse pointer, and
+  /// prevents this object's siblings (or any other objects that are not
```
```suggestion
  /// prevent this object's siblings (or any other objects that are not
```

gspencergoog commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5832,6 +5835,22 @@ class MouseRegion extends SingleChildRenderObjectWidget {
+  /// from detecting the pointer, thus affecting how their [onHover], [onEnter],
+  /// and [onExit] behave.
+  ///
+  /// If [opaque] is true, this widget will absorb the mouse pointer, and
```
```suggestion
  /// If [opaque] is true, this widget will absorb the mouse pointer and
```

gspencergoog commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5832,6 +5835,22 @@ class MouseRegion extends SingleChildRenderObjectWidget {
+  /// and [onExit] behave.
+  ///
+  /// If [opaque] is true, this widget will absorb the mouse pointer, and
+  /// prevents this widget's siblings (or any other widgets that are not
```
```suggestion
  /// prevent this widget's siblings (or any other widgets that are not
```

gspencergoog commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5832,6 +5835,22 @@ class MouseRegion extends SingleChildRenderObjectWidget {
+  ///
+  /// If [opaque] is true, this widget will absorb the mouse pointer, and
+  /// prevents this widget's siblings (or any other widgets that are not
+  /// ancestors or descendants of this widgets) from detecting the mouse
```
```suggestion
  /// ancestors or descendants of this widget) from detecting the mouse
```
or
```suggestion
  /// ancestors or descendants of these widgets) from detecting the mouse
```
depending on meaning.

gspencergoog reviewed (approved):
These are all just simple doc comments.  Code looks good.
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

dkwingsmt committed:
Merge remote-tracking branch 'upstream/master' into layer-hit-test

dkwingsmt committed:
Doc fixes
